Add a New Darktrace Solution

JustasZDT opened the pull request:
Adding an initial PR for a new Darktrace Solution. We wanted to rename the solution to more accurately reflect product naming conventions. This is looking to eventually supersede "AI Analyst Darktrace" Solution which is already published.

We have also changed the data ingestion method adding a new Data Connector which utilises Azure Monitor API. We POST events into the Sentinel API from our side and ingest these into a Custom Log Schema.

Alongside that there's a completely rebuilt workbook to reflect the new data schema.

Finally, we have added some analytics rules to convert the alerts and incidents we send through into Sentinel alerts and incidents respectively.

Hoping to get some initial feedback and run tests against this solution.

v-sabiraj commented:
Hi @JustasZDT, please sign the [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/Azure-Sentinel?pullRequest=5406) CLA, thanks.

v-sabiraj commented:
@JustasZDT, the name 'darktrace_model_alerts_CL' does not refer to any known table, tabular variable or function. Please add sample data, thanks.

v-sabiraj commented:
@JustasZDT, please add connectorId:'DarktraceRESTConnector' in 'ValidConnectorIds.json' file. Thanks.

v-sabiraj commented:
Hi @JustasZDT, can you please move sample data files to [Sample data folder](https://github.com/Azure/Azure-Sentinel/tree/master/Sample%20Data), thanks.

JustasZDT commented:
Hey, @v-sabiraj, thanks for that we've moved it now. I don't know whether you can manually re-trigger the pipeline, seems like something failed that didn't have anything to do with our content.

I have also moved out the Sample Data to the folder you referenced. We used the naming convention referenced in the contribution guidance, however I can see that most data in that folder is named in a different format - let me know if that needs to be renamed.

Finally, I've submitted a logo for light backgrounds, we have a different one for dark backgrounds but I'm not seeing anything in the contribution guidelines about that?

v-sabiraj commented on Solutions/Darktrace/Analytic Rules/CreateAlertFromSystemStatus.yaml:
```diff
@@ -0,0 +1,33 @@
+id: 2e629769-60eb-4a14-8bfc-bde9be66ebeb
+kind: scheduled
```
Kind should be "Scheduled" to "scheduled"

JustasZDT replied:
Capitalisation changed.

v-sabiraj commented on Solutions/Darktrace/Analytic Rules/CreateAlertFromModelBreach.yaml:
```diff
@@ -0,0 +1,75 @@
+id: a3c7b8ed-56a9-47b7-98e5-2555c16e17c9
+kind: scheduled
```
Kind should be "Scheduled" to "scheduled"

v-sabiraj commented:
@aprakash13, can you please review Analytic rules, thanks.

v-sabiraj commented on Solutions/Darktrace/Workbooks/DarktraceWorkbook.json:
```diff
@@ -0,0 +1,2628 @@
+            "type": 3,
+            "content": {
+              "version": "KqlItem/1.0",
+              "query": "darktrace_model_alerts_CL\r\n| where dtProduct_s ==\"Policy Breach\" and (modelName_s contains (\"SaaS\") or modelName_s contains (\"IaaS\")) and compliance_s == true\r\n| make-series count() default=0 on TimeGenerated in range({Timeframe:start}, now(), {Timeframe:grain})",
```
@JustasZDT,  compliance_s should be  compliance_b, please check and update, thanks.

seanmacdonald8 replied:
This has been updated now.

v-sabiraj commented on Solutions/Darktrace/Workbooks/DarktraceWorkbook.json:
```diff
@@ -0,0 +1,2628 @@
+            "type": 3,
+            "content": {
+              "version": "KqlItem/1.0",
+              "query": "{aia_type}",
```
@JustasZDT, please check and update the query, thanks.

seanmacdonald8 replied:
@v-sabiraj The KQL for that query is embedded within the selected aia_type, so it shouldn't need updating. We followed this sample online from a Microsoft form: https://techcommunity.microsoft.com/t5/azure-observability/if-statment-in-a-kql-query/m-p/1363551

v-sabiraj commented:
@JustasZDT, please remove sample data folder added in solution folder, Solutions/Darktrace/Sample Data/

v-sabiraj commented:
@JustasZDT, please rename Darktrace_RawLogs.json to darktrace_model_alerts_CL.json as queries are using this, thanks.

v-sabiraj commented:
@JustasZDT, the workbook content is added at both the places, in solution as well as stanalone workbook folder, please remove from standalone folder if not needed, thanks.

v-sabiraj commented on Solutions/Darktrace/Data Connectors/Doc_DarktraceConnectorRESTAPI.md:
```diff
@@ -0,0 +1,39 @@
+<!-- Taken from https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/DataConnectors/Templates/Doc_Template_REST_API_Connector.md -->
+
+# Connect your Darktrace Enteprise Immune System to Microsoft Sentinel 
```
@JustasZDT, please fix the typo Enteprise to Enterprise

seanmacdonald8 replied:
Typo has been fixed now.

v-spadarthi commented:
@JustasZDT : Could you please resolve below comments 
In workbook Darktrace AI Analyst section we are getting error like below
<img width="473" alt="image" src="https://user-images.githubusercontent.com/101796244/180966110-3cbbe669-ead9-443a-8907-b8cb2518fcfe.png">
I can see the workbook content in both the places like Solutions/Darktrace/Workbooks/Images/Preview and Workbooks/Images/Preview.
Can you please remove workbook from standalone folder(Workbooks/Images/Preview) if not needed.
While checking Analytics for [Solutions/Darktrace/Analytic Rules/CreateIncidentFromAIAnalystIncident.yaml] this we got an error like below we couldn't find groupScore_d and groupCategory_s columns in darktrace_model_alerts_CL table after ingesting data
<img width="497" alt="image" src="https://user-images.githubusercontent.com/101796244/180985829-825b9fe3-b55a-406f-bf40-104227c757ee.png">

v-spadarthi commented:
@aprakash13, can you please review Analytic rules, thanks.

JustasZDT commented:
Thanks, @v-spadarthi, we'll try to replicate on our side and @seanmacdonald8 will get back to you with feedback.

>I can see the workbook content in both the places like Solutions/Darktrace/Workbooks/Images/Preview and Workbooks/Images/Preview.
>Can you please remove workbook from standalone folder(Workbooks/Images/Preview) if not needed.

Less sure about this. The workbook .json is only present in our Solution folder. The Workbook preview images were added to the standalone folder because the CI pipeline was not detecting preview images that were placed in our Solution folder.

JustasZDT commented:
@v-spadarthi we have discovered that groupScore was missing from the test data which was generating errors. This should be resolved now with e3f7a9c8494d7494a1ad88c7c9aab338d14d05c9 so feel free to re-test and let us know what you think.

v-spadarthi commented:
@JustasZDT : Thanks for adding groupscore to test data.
We tested workbook is fine but analytics getting an error like below,
<img width="518" alt="image" src="https://user-images.githubusercontent.com/101796244/181479884-5a8266ec-f09d-48a0-b900-e588b22327ef.png">
Actually we requested 2  columns groupScore_d and groupCategory_s  not there in darktrace_model_alerts_CL table.But you added in test data only one ,can you please add other one also

aprakash13 reviewed (commented):
Queries look good. Please feel free to merge if other components look good.
Just one quick suggestion. Please move the '**kind**' field in all the queries to the end of the template for uniformity's sake similar to this example: https://github.com/Azure/Azure-Sentinel/blob/master/Detections/AuditLogs/AccountCreatedDeletedByNonApprovedUser.yaml

JustasZDT commented:
@aprakash13 noted, thanks, we'll move that around.

@v-spadarthi we'll add that to the test data, apologies.

To both - can I reiterate my question as I've not received an answer earlier - I've submitted a logo for light backgrounds, we have a different one for dark backgrounds but I'm not seeing anything in the contribution guidelines about that? Is the logo always placed in a white container?

JustasZDT commented:
Hello, we've made some more changes in line with the latest feedback we've received - looking forward to any further notes from yourselves. Also, reiterating the question asked above regarding logo colours.

v-spadarthi commented:
@JustasZDT : Thanks for providing the updated sample data now all are working fine.
<img width="935" alt="image" src="https://user-images.githubusercontent.com/101796244/182379842-8518ed4f-bc16-421d-9381-b489379a3cb0.png">
 Is the logo always placed in a white container?--There is no such criteria, you can use whatever you like but, Logo needs to be in SVG format and under 5 Kb and Logo scales well to fit in a 75 px square.

JustasZDT commented:
Lovely, thanks for reviewing @v-spadarthi. Are there any other components pending review or is this ready to merge?

JustasZDT commented:
Hey all, quick question - I've noticed that we have not added a solution package here, having serious problems with running the packaging tool. Is there somebody we can chat about this with?

JustasZDT commented:
Adding some context, running V1 and V2 packaging tools with the input files for both attached. The output is in collapsible sections for formatting.


## Running packaging tool V1

Input file:

```json
/**
 * Solution Automation Input File Json
 * -----------------------------------------------------
 * The purpose of this json is to provide detail on the various fields the input file can have.
 * Name: Solution Name - Ex. "Symantec Endpoint Protection"
 * Author: Author Name+Email of Solution - Ex. "Eli Forbes - v-eliforbes@microsoft.com"
 * Logo: Link to the Logo used in createUiDefinition.json
 * - NOTE: This field is only recommended for Azure Global Cloud. It is not recommended for solutions in Azure Government Cloud as the image will not be shown properly.
 * Description: Solution Description used in createUiDefinition.json. Can include markdown.
 * WorkbookDescription: Workbook description(s), generally from Workbooks' Metadata. This field can be a string if 1 description is used across all, and an array if multiple are used.
 * PlaybookDescription: Playbook description(s), generally from Playbooks' Metadata. This field can be a string if 1 description is used across all, and an array if multiple are used.
 * WatchlistDescription: Watchlist description(s), generally from Watchlists' Property data. This field can be a string if 1 description is used across all, and an array if multiple are used. This field is used if the description from the Watchlist resource is not desired in the Create-UI.
 * Workbooks, Analytic Rules, Playbooks, etc.: These fields take arrays of paths relative to the repo  root, or BasePath if provided.
 * SavedSearches: This input assumes a format of any of the following:
 * -- Direct export via API (see https://docs.microsoft.com/rest/api/loganalytics/saved-searches/list-by-workspace)
 * -- Array of SavedSearch resources
 * -- Raw ARM template
 *
 * - NOTE: Playbooks field can take standard Playbooks, Custom Connectors, and Function Apps
 * BasePath: Optional base path to use. Either Internet URL or File Path. Default is repo root (https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/)
 * Version: Version to be used during package creation. We should use any version >= 2.0.0 in case solution needs to be packaged for Template Spec
 * Metadata: Name of metadata file for the Solution, path is to be considered from BasePath.
 * TemplateSpec: Boolean value used to determine whether the package should be generated as a template spec
 */
 {
    "Name": "Darktrace",
    "Author": "Darktrace - info@darktrace.com",
    "Logo": "<img src=\"https://github.com/Azure/Azure-Sentinel/blob/master/Logos/Darktrace.svg\" width=\"75px\" height=\"75px\">",
    "Description": "Hello",
    "Workbooks": [
        "Workbooks/DarktraceWorkbook.json"
    ],
    "WorkbookDescription": "The Darktrace Workbook visualises AI Analyst, Model Breach and Antigena Email data received by the Darktrace Data Connector.",
    "Analytic Rules": [
        "Analytic Rules/CreateAlertFromModelBreach.yaml",
        "Analytic Rules/CreateAlertFromSystemStatus.yaml",
        "Analytic Rules/CreateIncidentFromAIAnalystIncident.yaml"    
    ],
    "Playbooks": [],
    "PlaybookDescription": ["{Description of playbook}"],
    "Parsers": [],
    "SavedSearches": [],
    "Hunting Queries": [],
    "Data Connectors": [
        "Data Connectors/DarktraceConnectorRESTAPI.json"
    ],
    "Watchlists": [],
    "WatchlistDescription": [],
    "BasePath": "https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Solutions/Darktrace/",
    "Version": "1.0.0",
    "Metadata": "SolutionMetadata.json",
    "TemplateSpec": false
  }
```

<details><summary>Packaging tool output</summary>

```powershell
PS C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution> .\createSolution.ps1
.\createSolution.ps1: File C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution\createSolution.ps1 cannot be loaded. The file C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution\createSolution.ps1 is not digitally signed. You cannot run this script on the current system. For more information about running scripts and setting execution policy, see about_Execution_Policies at https://go.microsoft.com/fwlink/?LinkID=135170.
PS C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution> Set-ExecutionPolicy Bypass
PS C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution> .\createSolution.ps1
Downloading Workbooks/DarktraceWorkbook.json
Generating Workbook using Workbooks/DarktraceWorkbook.json
Downloading Analytic Rules/CreateAlertFromModelBreach.yaml
Generating Alert Rule using Analytic Rules/CreateAlertFromModelBreach.yaml
Downloading Analytic Rules/CreateAlertFromSystemStatus.yaml
Generating Alert Rule using Analytic Rules/CreateAlertFromSystemStatus.yaml
Downloading Analytic Rules/CreateIncidentFromAIAnalystIncident.yaml
Generating Alert Rule using Analytic Rules/CreateIncidentFromAIAnalystIncident.yaml
Downloading {Description of playbook}
Failed to download {Description of playbook} -- Please ensure that it exists in https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Solutions/Darktrace//
Downloading Data Connectors/DarktraceConnectorRESTAPI.json
Generating Data Connector using Data Connectors/DarktraceConnectorRESTAPI.json
Removing empty property title
Removing empty property instructions
Downloading Data Connectors/DarktraceConnectorRESTAPI.json
fatal: not a git repository (or any of the parent directories): .git
Missing arm-ttk validations. Downloading module...

    Directory: C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          08/08/2022    17:42                tmp
VERBOSE: GET with 0-byte payload
VERBOSE: received 462521-byte response of content type application/x-zip-compressed
Get-ChildItem: C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution\tmp\arm-ttk\Expand-AzTemplate.ps1:123
Line |
 123 |                          Get-ChildItem -Filter *.json |
     |                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot find path 'C:\Solutions\Templates' because it does not exist.

Expand-AzTemplate: C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution\tmp\arm-ttk\Test-AzTemplate.ps1:441
Line |
 441 |  …      $expandedTemplate =Expand-AzTemplate -TemplatePath $templatePath
     |                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | No azuredeploy.json or mainTemplate.json found beneath /Solutions/Templates


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Allowed Values Should Actually Be Allowed
Timespan  : 00:00:00.1666420
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Controls In Outputs Must Exist
Timespan  : 00:00:00.0596845
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : CreateUIDefinition Must Not Have Blanks
Timespan  : 00:00:00.0218065
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : CreateUIDefinition Should Have Schema
Timespan  : 00:00:00.0397189
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Credential Confirmation Should Not Be Hidden
Timespan  : 00:00:00.0552032
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Handler Must Be Correct
Timespan  : 00:00:00.0189409
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : HideExisting Must Be Correctly Handled
Timespan  : 00:00:00.0639916
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Location Should Be In Outputs
Timespan  : 00:00:00.0201517
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Outputs Must Be Present In Template Parameters
Timespan  : 00:00:00.0168335
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Parameters Without Default Must Exist In CreateUIDefinition
Timespan  : 00:00:00.0152106
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Password Textboxes Must Be Used For Password Parameters
Timespan  : 00:00:00.0300372
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : PasswordBoxes Must Have Min Length
Timespan  : 00:00:00.0355756
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Textboxes Are Well Formed
Timespan  : 00:00:00.0357698
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Tooltips Should Be Present
Timespan  : 00:00:00.0485166
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Usernames Should Not Have A Default
Timespan  : 00:00:00.0374712
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Validations Must Have Message
Timespan  : 00:00:00.0528504
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : VMSizes Must Match Template
Timespan  : 00:00:00.0448163
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : adminUsername Should Not Be A Literal
Timespan  : 00:00:00.0506171
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : apiVersions Should Be Recent
Timespan  : 00:00:00.0642114
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : artifacts parameter
Timespan  : 00:00:00.0172607
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : CommandToExecute Must Use ProtectedSettings For Secrets
Timespan  : 00:00:00.0569269
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : DependsOn Best Practices
Timespan  : 00:00:00.0439622
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Deployment Resources Must Not Be Debug
Timespan  : 00:00:00.0406274
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : DeploymentTemplate Must Not Contain Hardcoded Uri
Timespan  : 00:00:00.0343642
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : DeploymentTemplate Schema Is Correct
Timespan  : 00:00:00.0143931
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Dynamic Variable References Should Not Use Concat
Timespan  : 00:00:00.0177567
File      : {Name, Extension, Text, FullPath…}


Errors    : {Property: "EventID" must use one of the following expressions for an resourceId property:
                        extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,
            variables,subscription,guid, Property: "CurrentGroupID" must use one of the following expressions for an
            resourceId property:
                        extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,
            variables,subscription,guid}
Warnings  : {}
Output    : {}
AllOutput : {Property: "EventID" must use one of the following expressions for an resourceId property:
                        extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,
            variables,subscription,guid, Property: "CurrentGroupID" must use one of the following expressions for an
            resourceId property:
                        extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,
            variables,subscription,guid}
Passed    : False
Group     : deploymentTemplate
Name      : IDs Should Be Derived From ResourceIDs
Timespan  : 00:00:00.1016589
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Location Should Not Be Hardcoded
Timespan  : 00:00:00.0278243
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : ManagedIdentityExtension must not be used
Timespan  : 00:00:00.0147100
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Min And Max Value Are Numbers
Timespan  : 00:00:00.0140927
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Outputs Must Not Contain Secrets
Timespan  : 00:00:00.0171864
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Parameters Must Be Referenced
Timespan  : 00:00:00.1521975
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : providers apiVersions Is Not Permitted
Timespan  : 00:00:00.0678172
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : ResourceIds should not contain
Timespan  : 00:00:00.0591986
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Resources Should Have Location
Timespan  : 00:00:00.0185881
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Resources Should Not Be Ambiguous
Timespan  : 00:00:00.0262558
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Secure String Parameters Cannot Have Default
Timespan  : 00:00:00.0152015
File      : {Name, Extension, Text, FullPath…}


Errors    : {Empty property:  null found on line: 170, Empty property:  null found on line: 201, Empty property:  null
            found on line: 203, Empty property:  null found on line: 253…}
Warnings  : {}
Output    : {}
AllOutput : {Empty property:  null found on line: 170, Empty property:  null found on line: 201, Empty property:  null
            found on line: 203, Empty property:  null found on line: 253…}
Passed    : False
Group     : deploymentTemplate
Name      : Template Should Not Contain Blanks
Timespan  : 00:00:00.0725049
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Variables Must Be Referenced
Timespan  : 00:00:00.0387856
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Virtual Machines Should Not Be Preview
Timespan  : 00:00:00.0923657
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : VM Images Should Use Latest Version
Timespan  : 00:00:00.0206460
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : VM Size Should Be A Parameter
Timespan  : 00:00:00.0443680
File      : {Name, Extension, Text, FullPath…}

Failed arm-ttk (Test-AzTemplate): Package
Failed arm-ttk (Test-AzTemplate) on solutions: Package
```

</details>


## Running package tool V2

**Input file:**

```json
  /**
 * Solution Automation Input File Json
 * -----------------------------------------------------
 * The purpose of this json is to provide detail on the various fields the input file can have.
 * Name: Solution Name - Ex. "Symantec Endpoint Protection"
 * Author: Author Name+Email of Solution - Ex. "Eli Forbes - v-eliforbes@microsoft.com"
 * Logo: Link to the Logo used in createUiDefinition.json
 * - NOTE: This field is only recommended for Azure Global Cloud. It is not recommended for solutions in Azure Government Cloud as the image will not be shown properly.
 * Description: Solution Description used in createUiDefinition.json. Can include markdown.
 * WorkbookDescription: Workbook description(s), generally from Workbooks' Metadata. This field can be a string if 1 description is used across all, and an array if multiple are used.
 * PlaybookDescription: Playbook description(s), generally from Playbooks' Metadata. This field can be a string if 1 description is used across all, and an array if multiple are used.
 * WatchlistDescription: Watchlist description(s), generally from Watchlists' Property data. This field can be a string if 1 description is used across all, and an array if multiple are used. This field is used if the description from the Watchlist resource is not desired in the Create-UI.
 * Workbooks, Analytic Rules, Playbooks, etc.: These fields take arrays of paths relative to the repo  root, or BasePath if provided.
 * SavedSearches: This input assumes a format of any of the following:
 * -- Direct export via API (see https://docs.microsoft.com/rest/api/loganalytics/saved-searches/list-by-workspace)
 * -- Array of SavedSearch resources
 * -- Raw ARM template
 *
 * - NOTE: Playbooks field can take standard Playbooks, Custom Connectors, and Function Apps
 * BasePath: Optional base path to use. Either Internet URL or File Path. Default is repo root (https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/)
 * Version: Version to be used during package creation. We should use any version >= 2.0.0 in case solution needs to be packaged for Template Spec
 * Metadata: Name of metadata file for the Solution, path is to be considered from BasePath.
 * TemplateSpec: Boolean value used to determine whether the package should be generated as a template spec
 */
{
    "Name": "Darktrace",
    "Author": "Darktrace - info@darktrace.com",
    "Logo": "<img src=\"https://github.com/Azure/Azure-Sentinel/blob/master/Logos/Darktrace.svg\" width=\"75px\" height=\"75px\">",
    "Description": "Hello",
    "WorkbookDescription": "The Darktrace Workbook visualises AI Analyst, Model Breach and Antigena Email data received by the Darktrace Data Connector.",
    "Workbooks": [
        "Workbooks/DarktraceWorkbook.json"
    ],
    "WorkbookDescription": "The Darktrace Workbook visualises AI Analyst, Model Breach and Antigena Email data received by the Darktrace Data Connector.",
    "AnalyticalRuleBladeDescription": "The Darktrace solution installs Analytical Rules in order to create incidents and alerts on Sentinel based on Darktrace data.",
    "HuntingQueryBladeDescription": "",
    "PlaybooksBladeDescription": "",
    "Analytic Rules": [
        "Analytic Rules/CreateAlertFromModelBreach.yaml",
        "Analytic Rules/CreateAlertFromSystemStatus.yaml",
        "Analytic Rules/CreateIncidentFromAIAnalystIncident.yaml"    
    ],
    "Playbooks": [],
    "PlaybookDescription": [""],
    "Parsers": [],
    "SavedSearches": [],
    "Hunting Queries": [],
    "Data Connectors": [],
    "Watchlists": [],
    "WatchlistDescription": [],
    "BasePath": "https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Solutions/Darktrace/",
    "Version": "1.0.0",
    "Metadata": "SolutionMetadata.json",
    "TemplateSpec": true,
    "Is1PConnector": false
  }
```


<details><summary>Output</summary>

```powershell
PS C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution\V2> .\createSolutionV2.ps1
Downloading https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Solutions/Darktrace//Workbooks/DarktraceWorkbook.json
Generating Workbook using Workbooks/DarktraceWorkbook.json
Downloading https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Tools/Create-Azure-Sentinel-Solution/V2/WorkbookMetadata/WorkbooksMetadata.json
Downloading https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Solutions/Darktrace//Analytic Rules/CreateAlertFromModelBreach.yaml
Generating Alert Rule using Analytic Rules/CreateAlertFromModelBreach.yaml
Downloading https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Solutions/Darktrace//Analytic Rules/CreateAlertFromSystemStatus.yaml
Generating Alert Rule using Analytic Rules/CreateAlertFromSystemStatus.yaml
Downloading https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Solutions/Darktrace//Analytic Rules/CreateIncidentFromAIAnalystIncident.yaml
Generating Alert Rule using Analytic Rules/CreateIncidentFromAIAnalystIncident.yaml
Downloading https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Solutions/Darktrace//
Failed to download https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Solutions/Darktrace// -- Please ensure that it exists in https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Solutions/Darktrace//
Downloading C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution\V2/../../../Solutions/Darktrace/SolutionMetadata.json
fatal: not a git repository (or any of the parent directories): .git
Missing arm-ttk validations. Downloading module...

    Directory: C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          08/08/2022    17:42                tmp
VERBOSE: GET with 0-byte payload
VERBOSE: received 462521-byte response of content type application/x-zip-compressed
Get-ChildItem: C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution\tmp\arm-ttk\Expand-AzTemplate.ps1:123
Line |
 123 |                          Get-ChildItem -Filter *.json |
     |                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot find path 'C:\Solutions\Templates' because it does not exist.

Expand-AzTemplate: C:\Azure-Sentinel\Azure-Sentinel-master\Tools\Create-Azure-Sentinel-Solution\tmp\arm-ttk\Test-AzTemplate.ps1:441
Line |
 441 |  …      $expandedTemplate =Expand-AzTemplate -TemplatePath $templatePath
     |                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | No azuredeploy.json or mainTemplate.json found beneath /Solutions/Templates


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Allowed Values Should Actually Be Allowed
Timespan  : 00:00:00.2662878
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Controls In Outputs Must Exist
Timespan  : 00:00:00.0491586
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : CreateUIDefinition Must Not Have Blanks
Timespan  : 00:00:00.0240897
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : CreateUIDefinition Should Have Schema
Timespan  : 00:00:00.0618814
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Credential Confirmation Should Not Be Hidden
Timespan  : 00:00:00.0884655
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Handler Must Be Correct
Timespan  : 00:00:00.0171269
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : HideExisting Must Be Correctly Handled
Timespan  : 00:00:00.0915351
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Location Should Be In Outputs
Timespan  : 00:00:00.0190531
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Outputs Must Be Present In Template Parameters
Timespan  : 00:00:00.0218276
File      : {Name, Extension, Text, FullPath…}


Errors    : {workbook1-name does not have a default value, and is not defined in CreateUIDefinition.outuputs}
Warnings  : {}
Output    : {}
AllOutput : workbook1-name does not have a default value, and is not defined in CreateUIDefinition.outuputs
Passed    : False
Group     : CreateUIDefinition
Name      : Parameters Without Default Must Exist In CreateUIDefinition
Timespan  : 00:00:00.0193071
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Password Textboxes Must Be Used For Password Parameters
Timespan  : 00:00:00.0411668
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : PasswordBoxes Must Have Min Length
Timespan  : 00:00:00.0390263
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Textboxes Are Well Formed
Timespan  : 00:00:00.0491547
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Tooltips Should Be Present
Timespan  : 00:00:00.0593738
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Usernames Should Not Have A Default
Timespan  : 00:00:00.0353231
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : Validations Must Have Message
Timespan  : 00:00:00.0782936
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : CreateUIDefinition
Name      : VMSizes Must Match Template
Timespan  : 00:00:00.0668912
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : adminUsername Should Not Be A Literal
Timespan  : 00:00:00.1141084
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : apiVersions Should Be Recent
Timespan  : 00:00:00.1320470
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : artifacts parameter
Timespan  : 00:00:00.0357421
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : CommandToExecute Must Use ProtectedSettings For Secrets
Timespan  : 00:00:00.0997527
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : DependsOn Best Practices
Timespan  : 00:00:00.0908419
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Deployment Resources Must Not Be Debug
Timespan  : 00:00:00.1017997
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : DeploymentTemplate Must Not Contain Hardcoded Uri
Timespan  : 00:00:00.0558590
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : DeploymentTemplate Schema Is Correct
Timespan  : 00:00:00.0207590
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Dynamic Variable References Should Not Use Concat
Timespan  : 00:00:00.0186226
File      : {Name, Extension, Text, FullPath…}


Errors    : {Property: "connectorId" must use one of the following expressions for an resourceId property:
                        extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,
            variables,subscription,guid, Property: "EventID" must use one of the following expressions for an
            resourceId property:
                        extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,
            variables,subscription,guid, Property: "connectorId" must use one of the following expressions for an
            resourceId property:
                        extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,
            variables,subscription,guid, Property: "connectorId" must use one of the following expressions for an
            resourceId property:
                        extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,
            variables,subscription,guid…}
Warnings  : {}
Output    : {}
AllOutput : {Property: "connectorId" must use one of the following expressions for an resourceId property:
                        extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,
            variables,subscription,guid, Property: "EventID" must use one of the following expressions for an
            resourceId property:
                        extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,
            variables,subscription,guid, Property: "connectorId" must use one of the following expressions for an
            resourceId property:
                        extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,
            variables,subscription,guid, Property: "connectorId" must use one of the following expressions for an
            resourceId property:
                        extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,
            variables,subscription,guid…}
Passed    : False
Group     : deploymentTemplate
Name      : IDs Should Be Derived From ResourceIDs
Timespan  : 00:00:00.1624575
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Location Should Not Be Hardcoded
Timespan  : 00:00:00.0590251
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : ManagedIdentityExtension must not be used
Timespan  : 00:00:00.0189901
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Min And Max Value Are Numbers
Timespan  : 00:00:00.0171951
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Outputs Must Not Contain Secrets
Timespan  : 00:00:00.0214623
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Parameters Must Be Referenced
Timespan  : 00:00:00.1651358
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : providers apiVersions Is Not Permitted
Timespan  : 00:00:00.0552417
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : ResourceIds should not contain
Timespan  : 00:00:00.0259564
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Resources Should Have Location
Timespan  : 00:00:00.0219976
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Resources Should Not Be Ambiguous
Timespan  : 00:00:00.0335443
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Secure String Parameters Cannot Have Default
Timespan  : 00:00:00.0561334
File      : {Name, Extension, Text, FullPath…}


Errors    : {Empty property:  "" found on line: 45, Empty property:  "" found on line: 46, Empty property:  "" found
            on line: 109}
Warnings  : {}
Output    : {}
AllOutput : {Empty property:  "" found on line: 45, Empty property:  "" found on line: 46, Empty property:  "" found
            on line: 109}
Passed    : False
Group     : deploymentTemplate
Name      : Template Should Not Contain Blanks
Timespan  : 00:00:00.0812944
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Variables Must Be Referenced
Timespan  : 00:00:00.0878444
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : Virtual Machines Should Not Be Preview
Timespan  : 00:00:00.1829605
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : VM Images Should Use Latest Version
Timespan  : 00:00:00.0189469
File      : {Name, Extension, Text, FullPath…}


Errors    : {}
Warnings  : {}
Output    : {}
AllOutput :
Passed    : True
Group     : deploymentTemplate
Name      : VM Size Should Be A Parameter
Timespan  : 00:00:00.1015714
File      : {Name, Extension, Text, FullPath…}

Failed arm-ttk (Test-AzTemplate): Package
Failed arm-ttk (Test-AzTemplate) on solutions: Package
```


</details>

NikTripathi commented:
Hi @JustasZDT , @v-sabiraj Will connect with you over teams. You can discuss issue with him. 
Thanks
